File: pb_pay_delivery/static/src/js/pb_pay_delivery.js
/** @odoo-module **/
/**
 * Pay & Deliver — a bespoke full-screen experience launched from the Pay Runs
 * cockpit (or the sidebar, which shows a run picker first).
 *   · Bank file       — prepare it, follow it, download the approved one.
 *   · Payment release — ask for the money to be sent; see it land.
 *   · Payslips out    — password-PDF batch delivery with per-slip status pills.
 *
 * PHASE 5. Every lane used to be a button that did the thing. Each is now a
 * button that ASKS — and where the business published "no approval needed" the
 * button says so on its face, so nobody is ever surprised by what a press
 * costs. A lane that is waiting names the person it is waiting for and offers
 * the door to the request: waiting is never a dead end.
 *
 * RPC facade: pb.pay.delivery. pbim-tokenized (.pbpd.pbim). Lucide icons only.
 */
import { Component, useState, onWillStart, onMounted, onWillUnmount } from "@odoo/owl";
import { registry } from "@web/core/registry";
import { useService } from "@web/core/utils/hooks";
import { _t } from "@web/core/l10n/translation";
import { ic } from "@pb_import_kit/js/import_icons";

const MODEL = "pb.pay.delivery";

export class PbPayDelivery extends Component {
    static template = "pb_pay_delivery.PbPayDelivery";
    static props = { action: { type: Object, optional: true }, "*": true };

    setup() {
        this.orm = useService("orm");
        this.notif = useService("notification");
        this.action = useService("action");
        this.ic = ic;
        const runId = this.props.action?.params?.run_id;
        // Somewhere else sent the person here and has to be able to take them
        // back. Deliberately read from the action's own context rather than
        // imported from the hub: this cockpit does not depend on that module,
        // and an import it does not have takes the whole bundle down.
        this.back = this.props.action?.context?.pb_back || null;
        this.state = useState({
            loaded: false,
            busy: false,
            runId: runId || null,
            picking: !runId,     // no run → show the picker
            runs: [],
            data: null,
            bank: null,
            companyAccount: "",
            bankReference: "",
            showExcl: false,
            netShown: 0,         // count-up display
        });
        onWillStart(async () => {
            if (this.state.runId) { await this.load(); }
            else { await this.loadRuns(); }
        });
        onMounted(() => this._startCountUp());
        onWillUnmount(() => this._stopCountUp());
    }

    // --------------------------------------------------------------- loaders
    async loadRuns() {
        try {
            this.state.runs = await this.orm.call(MODEL, "get_recent_runs", []);
            this.state.loaded = true;
        } catch (e) {
            this.notif.add(this._err(e), { type: "danger" });
            this.state.loaded = true;
        }
    }

    /** Back to whatever sent us here, with everything it was holding. */
    goBack() {
        if (!this.back) { return; }
        const context = this.back.context || {};
        this.action.doAction({
            type: "ir.actions.client",
            tag: this.back.tag,
            name: this.back.label,
            target: "current",
            params: { ...context },
            context: { ...context },
        }, { clearBreadcrumbs: true });
    }

    async pickRun(id) {
        this.state.runId = id;
        this.state.picking = false;
        this.state.loaded = false;
        await this.load();
        this._startCountUp();
    }

    async load() {
        try {
            const d = await this.orm.call(MODEL, "get_delivery_data", [this.state.runId]);
            this.state.data = d;
            if (!this.state.bank && d.banks.length) {
                // default to the bank the last file used, else Vietcombank,
                // else the first layout there is
                const used = d.banks.find(b => b.key === d.bank_file.bank_format);
                const vcb = d.banks.find(b => b.key === "vietcombank");
                this.state.bank = (used || vcb || d.banks[0]).key;
            }
            if (d.release && d.release.bank_reference) {
                this.state.bankReference = d.release.bank_reference;
            }
            this.state.loaded = true;
        } catch (e) {
            this.notif.add(this._err(e), { type: "danger" });
            this.state.loaded = true;
        }
    }

    // --------------------------------------------------------------- count-up
    _startCountUp() {
        this._stopCountUp();
        const target = (this.state.data && this.state.data.run.total_net) || 0;
        if (!target) { this.state.netShown = 0; return; }
        const t0 = 900, start = performance.now();
        const step = (now) => {
            const p = Math.min(1, (now - start) / t0);
            const eased = 1 - Math.pow(1 - p, 3);
            this.state.netShown = Math.round(target * eased);
            if (p < 1) { this._raf = requestAnimationFrame(step); }
        };
        this._raf = requestAnimationFrame(step);
    }
    _stopCountUp() { if (this._raf) { cancelAnimationFrame(this._raf); this._raf = null; } }

    // --------------------------------------------------------------- bank file
    selectBank(key) { this.state.bank = key; }
    onCompanyAccount(ev) { this.state.companyAccount = ev.target.value; }
    onBankReference(ev) { this.state.bankReference = ev.target.value; }
    toggleExcl() { this.state.showExcl = !this.state.showExcl; }

    get bankFile() { return (this.state.data && this.state.data.bank_file) || {}; }
    get release() { return (this.state.data && this.state.data.release) || {}; }
    get delivery() { return (this.state.data && this.state.data.delivery) || {}; }

    /** What the bank-file button says, so a press never surprises anybody. */
    get prepareLabel() {
        const again = this.bankFile.exists;
        if (this.bankFile.fast_lane) {
            return again ? _t("Prepare again — approved on the spot")
                         : _t("Prepare and approve now (no approval needed)");
        }
        return again ? _t("Prepare it again and send it in")
                     : _t("Prepare bank file for approval");
    }

    get releaseLabel() {
        return this.release.fast_lane
            ? _t("Release the money now (no approval needed)")
            : _t("Send the release for approval");
    }

    get sendLabel() {
        return this.delivery.fast_lane
            ? _t("Send payslips")
            : _t("Send payslips for approval");
    }

    async prepareBankFile() {
        if (!this.state.bank || this.state.busy) { return; }
        this.state.busy = true;
        try {
            await this.orm.call(MODEL, "prepare_bank_file",
                [this.state.runId, this.state.bank, this.state.companyAccount || false]);
            await this.load();
            const f = this.bankFile;
            this.notif.add(
                f.state === "approved"
                    ? _t("The bank file is ready to download.")
                    : _t("The bank file is with %s for approval.",
                         f.with_whom || _t("its approver")),
                { type: "success" });
        } catch (e) {
            this.notif.add(this._err(e), { type: "danger" });
        } finally {
            this.state.busy = false;
        }
    }

    async downloadFile() {
        if (this.state.busy) { return; }
        this.state.busy = true;
        try {
            const act = await this.orm.call(MODEL, "download_bank_file", [this.state.runId]);
            await this.action.doAction(act);
            await this.load();
        } catch (e) {
            this.notif.add(this._err(e), { type: "danger" });
        } finally {
            this.state.busy = false;
        }
    }

    // --------------------------------------------------------------- release
    async requestRelease() {
        if (this.state.busy) { return; }
        this.state.busy = true;
        try {
            await this.orm.call(MODEL, "request_release",
                [this.state.runId, this.state.bankReference || false]);
            await this.load();
            const r = this.release;
            this.notif.add(
                r.state === "released"
                    ? _t("The money is marked as released.")
                    : _t("The release is with %s.", r.with_whom || _t("its approver")),
                { type: "success" });
        } catch (e) {
            this.notif.add(this._err(e), { type: "danger" });
        } finally {
            this.state.busy = false;
        }
    }

    // --------------------------------------------------------------- payslips
    async send(forceAll) {
        if (this.state.busy) { return; }
        this.state.busy = true;
        try {
            await this.orm.call(MODEL, "send_payslips", [this.state.runId, !!forceAll]);
            await this.load();
            const d = this.delivery;
            if (d.state === "pending") {
                this.notif.add(
                    _t("The payslips are with %s for approval.",
                       d.with_whom || _t("their approver")), { type: "success" });
            } else {
                this.notif.add(
                    _t("Delivery complete — %s sent, %s failed, %s skipped.",
                       d.sent, d.failed, d.skipped),
                    { type: d.failed ? "warning" : "success" });
            }
        } catch (e) {
            this.notif.add(this._err(e), { type: "danger" });
        } finally {
            this.state.busy = false;
        }
    }

    // --------------------------------------------------------------- the door
    async openRequest(requestId) {
        if (!requestId) { return; }
        const act = await this.orm.call(MODEL, "open_request", [requestId]);
        this.action.doAction(act);
    }

    backToRuns() {
        this.state.picking = true;
        this.state.data = null;
        this.state.runId = null;
        this.state.bank = null;
        this.loadRuns();
    }

    // --------------------------------------------------------------- helpers
    get bankName() {
        const b = this.state.data && this.state.data.banks.find(x => x.key === this.state.bank);
        return b ? b.name : "";
    }
    money(n) {
        if (n === null || n === undefined) { return "—"; }
        const cur = (this.state.data && this.state.data.currency) || "₫";
        return cur + " " + Math.round(n).toLocaleString("en-US");
    }
    fileSize(bytes) {
        if (!bytes) { return "0 B"; }
        if (bytes >= 1024 * 1024) { return (bytes / 1048576).toFixed(1) + " MB"; }
        if (bytes >= 1024) { return (bytes / 1024).toFixed(1) + " KB"; }
        return bytes + " B";
    }
    /** The tone of a lane's status chip. One vocabulary for all three. */
    toneOf(state) {
        return {
            approved: "ok", released: "ok", done: "ok", posted: "ok",
            pending: "wait", draft: "", returned: "warn", sending: "wait",
            rejected: "bad", superseded: "muted",
        }[state] || "";
    }
    stepTone(status) {
        return { done: "ok", active: "wait", blocked: "bad",
                 returned: "warn", skipped: "muted" }[status] || "";
    }
    pillClass(state) {
        return { sent: "ok", failed: "bad", skipped_no_email: "warn" }[state] || "";
    }
    pillLabel(state) {
        return { sent: _t("Sent"), failed: _t("Failed"),
                 skipped_no_email: _t("No email") }[state] || state;
    }
    reasonList(reasons) { return (reasons || []).join(" · "); }
    // deep-link an excluded row to the employee's record so the reviewer can go
    // straight to the bank account that failed validation (payload carries the id)
    openEmployeeBank(employeeId) {
        if (!employeeId) { return; }
        this.action.doAction({
            type: "ir.actions.act_window",
            res_model: "hr.employee",
            res_id: employeeId,
            views: [[false, "form"]],
            target: "current",
        });
    }
    _err(e) {
        return (e && e.data && e.data.message) || (e && e.message) || _t("Action failed.");
    }
}

registry.category("actions").add("pb_pay_delivery", PbPayDelivery);

File: pb_pay_delivery/models/pb_pay_delivery.py
# -*- coding: utf-8 -*-
"""RPC facade for the bespoke full-screen Pay & Deliver experience.

An AbstractModel (no table) exposing three lanes to the OWL client action:
  * bank file       — prepare, send in, follow, download the approved one,
  * payment release — ask for the money to be sent, follow, see it land,
  * payslips out    — recipient preview + encrypted-PDF batch delivery + log.

Access is enforced HERE (safety rail 3): an ir.actions.client carries no group
field, so every public method guards with `_require_access()` → AccessError.
The resolved PDF password is never returned by any method (rail 4).

WHAT PHASE 5 CHANGED. Every one of the three lanes used to be a button that
did the thing. Each is now a button that ASKS — and the answer, including "no
approval needed" where the business published that, is recorded. The payload
each lane returns therefore carries the same three things: what state it is in,
who it is with, and the door to the request.
"""

from odoo import _, api, fields, models
from odoo.exceptions import AccessError

from .approval_common import PAY_GROUPS as _PAY_GROUPS


class PbPayDelivery(models.AbstractModel):
    _name = 'pb.pay.delivery'
    _description = 'Pay & Deliver Cockpit'

    # ------------------------------------------------------------- access
    @api.model
    def _require_access(self):
        user = self.env.user
        if user._is_admin():
            return
        for g in _PAY_GROUPS:
            try:
                if user.has_group(g):
                    return
            except (ValueError, KeyError):
                continue
        raise AccessError(_(
            "You are not allowed to open Pay & Deliver. This requires the "
            "Payroll Manager or a Finance role."))

    # ------------------------------------------------------------- helpers
    def _run(self, run_id):
        return self.env['hr.payslip.run'].browse(int(run_id))

    def _done_slips(self, run):
        return run.slip_ids.filtered(lambda s: s.state == 'done')

    def _wizard(self, run):
        return self.env['vietnam.bank.export.wizard'].create({
            'payslip_run_id': run.id})

    def _validation_preview(self, run):
        """Format-independent row validation (account/holder/bank/net)."""
        wiz = self._wizard(run)
        registry = self.env['pb.bank.registry']
        eligible, excluded, total_net = 0, [], 0.0
        for slip in self._done_slips(run).sorted(key=lambda s: s.employee_id.name or ''):
            sources, reasons, net = wiz._row_sources(slip, eligible + 1, registry)
            if reasons:
                excluded.append({
                    'employee': slip.employee_id.name or '',
                    'employee_id': slip.employee_id.id,
                    'reasons': reasons,
                })
            else:
                eligible += 1
                total_net += net
        return eligible, excluded, total_net

    def _batch(self, run, create=False):
        batch = self.env['pb.payslip.delivery.batch'].search(
            [('run_id', '=', run.id)], order='create_date desc', limit=1)
        if not batch and create:
            batch = self.env['pb.payslip.delivery.batch'].create({
                'run_id': run.id, 'name': _('Delivery — %s', run.name)})
        return batch

    # --------------------------------------------------- the approval payload
    def _route_payload(self, record):
        """The same three answers for every lane: state, who, the door.

        `pending` is not enough on its own — an approver's name is what turns
        "waiting" from a dead end into something a person can chase.
        """
        if not record:
            return {'exists': False, 'state': '', 'state_label': '',
                    'request_id': 0, 'with_whom': '', 'block_reason': '',
                    'route': []}
        request = record.approval_request_id
        steps = []
        if request:
            for step in request.step_ids.sorted('sequence'):
                if not step.included or step.kind == 'fast':
                    continue
                steps.append({
                    'title': step.title,
                    'status': step.status,
                    'people': sorted({seat.acting_user_id.name or ''
                                      for seat in step.seat_ids}),
                })
        return {
            'exists': True,
            'id': record.id,
            'state': record.state,
            'state_label': dict(
                record._fields['state'].selection).get(record.state, ''),
            'request_id': request.id if request else 0,
            'with_whom': record._waiting_for(),
            'block_reason': request.block_reason or '' if request else '',
            'route': steps,
        }

    def _fast_lane(self, model, run):
        """Has the business said this one needs no approval?

        Asked of the ENGINE, every read, and never remembered: a route that
        exists exists whether or not anybody ticked a box, and the answer
        decides what the button says.
        """
        Model = self.env[model]
        process = self.env['biz.approval.process']._by_key(
            Model._approval_process_key)
        if not process:
            return False
        company = getattr(run, 'company_id', False) or self.env.company
        keys, _label = self.env['pb.bank.file'].scope_of_run(run)
        resolved = self.env['biz.approval.engine'].sudo().resolve_binding(
            company.id, process.key, keys, 'any')
        if resolved.get('error') or not resolved.get('version_id'):
            return False
        version = self.env['biz.approval.workflow.version'].sudo().browse(
            resolved['version_id'])
        from odoo.addons.biz_approval_workflow.models import definition as D
        steps = [s for s in D.normalise(version.definition)['steps']
                 if s['kind'] not in ('notify',)]
        return (not steps) or any(s['kind'] == 'fast' for s in steps)

    def _bank_file_payload(self, run):
        latest = self.env['pb.bank.file'].search(
            [('run_id', '=', run.id), ('state', '!=', 'superseded')],
            order='id desc', limit=1)
        payload = self._route_payload(latest)
        payload.update({
            'filename': latest.filename or '' if latest else '',
            'row_count': latest.row_count if latest else 0,
            'control_total': latest.control_total if latest else 0.0,
            'excluded_count': latest.excluded_count if latest else 0,
            'byte_size': latest.byte_size if latest else 0,
            'bank_format': latest.bank_format or '' if latest else '',
            'downloadable': bool(latest and latest.state == 'approved'),
            'fast_lane': self._fast_lane('pb.bank.file', run),
        })
        return payload

    def _release_payload(self, run):
        latest = self.env['pb.payment.release'].search(
            [('run_id', '=', run.id)], order='id desc', limit=1)
        payload = self._route_payload(latest)
        payload.update({
            'amount': latest.amount if latest else 0.0,
            'beneficiaries': latest.beneficiaries if latest else 0,
            'bank_reference': latest.bank_reference or '' if latest else '',
            'released_on': fields.Datetime.to_string(latest.released_on) or ''
            if latest else '',
            'fast_lane': self._fast_lane('pb.payment.release', run),
        })
        return payload

    def _delivery_payload(self, batch, run=None):
        base = {'exists': False, 'state': 'draft', 'sent': 0, 'failed': 0,
                'skipped': 0, 'lines': [], 'request_id': 0, 'with_whom': '',
                'state_label': '', 'route': [],
                'fast_lane': self._fast_lane('pb.payslip.delivery.batch', run)
                if run else True}
        if not batch:
            return base
        lines = [{
            'employee': l.employee_id.name or '',
            'email': l.email or '',
            'state': l.state,
            'error': l.error or '',
        } for l in batch.line_ids]
        base.update(self._route_payload(batch))
        base.update({
            'exists': True, 'state': batch.state,
            'sent': batch.sent_count, 'failed': batch.failed_count,
            'skipped': batch.skipped_count, 'lines': lines,
        })
        return base

    # ------------------------------------------------------------- run picker
    @api.model
    def get_recent_runs(self):
        """Recent pay runs (newest first) for the sidebar entry's picker.

        Only runs that actually have something to pay/deliver are offered — i.e.
        an approved run OR one carrying at least one confirmed (done) payslip.
        A draft/empty run has zero deliverable slips (``_done_slips`` filters on
        ``state == 'done'``), so listing it would only dead-end the picker."""
        self._require_access()
        runs = self.env['hr.payslip.run'].search(
            ['|', ('state', '=', 'done'), ('slip_ids.state', '=', 'done')],
            order='date_end desc, id desc', limit=24)
        out = []
        for run in runs:
            done = self._done_slips(run)
            out.append({
                'id': run.id, 'name': run.name, 'state': run.state,
                'period': "%s → %s" % (run.date_start, run.date_end)
                if run.date_start and run.date_end else '',
                'headcount': len(done),
                'paid': bool(getattr(run, 'pb_paid', False)),
            })
        return out

    # ------------------------------------------------------------- load
    @api.model
    def get_delivery_data(self, run_id):
        self._require_access()
        run = self._run(run_id)
        done = self._done_slips(run)

        eligible, excluded, total_net = self._validation_preview(run)

        # Recipient preview (payslips-out lane).
        recips, no_email = [], []
        for slip in done.sorted(key=lambda s: s.employee_id.name or ''):
            emp = slip.employee_id
            entry = {'employee': emp.name or '',
                     'avatar_url': '/web/image/hr.employee/%s/avatar_128' % emp.id,
                     'email': emp.work_email or ''}
            (recips if emp.work_email else no_email).append(entry)

        banks = [{'key': l.bank_format, 'name': l.name}
                 for l in self.env['pb.bank.file.layout'].search(
                     [], order='name')]

        currency = (self.env.company.currency_id.symbol or '₫')
        pattern = self.env['ir.config_parameter'].sudo().get_param(
            'pb_pay_delivery.pdf_password_pattern') or '{account_last4}{birth_year}'

        return {
            'run': {
                'id': run.id, 'name': run.name,
                'period': "%s → %s" % (run.date_start, run.date_end)
                if run.date_start and run.date_end else '',
                'headcount': len(done),
                'total_net': total_net,
                'state': run.state,
                'approved': (run.state or '') == 'done',
                'paid': bool(getattr(run, 'pb_paid', False)),
            },
            'currency': currency,
            'banks': banks,
            'validation': {
                'eligible': eligible,
                'excluded_count': len(excluded),
                'excluded': excluded,
            },
            'recipients': {
                'count': len(recips),
                'no_email_count': len(no_email),
                'avatars': [r['avatar_url'] for r in recips[:14]],
                'no_email': no_email,
            },
            'password_pattern': pattern,
            'bank_file': self._bank_file_payload(run),
            'release': self._release_payload(run),
            'delivery': self._delivery_payload(self._batch(run), run=run),
            'is_manager': True,
        }

    # ------------------------------------------------------------- money out
    @api.model
    def prepare_bank_file(self, run_id, bank_format, company_account=None):
        """Generate the file, store it, and send it in for approval."""
        self._require_access()
        run = self._run(run_id)
        record = self.env['pb.bank.file'].prepare(
            run, bank_format, company_account or None)
        self.env['biz.approval.engine'].submit(record)
        return self._bank_file_payload(run)

    @api.model
    def download_bank_file(self, run_id):
        """The approved bytes, and nothing else."""
        self._require_access()
        run = self._run(run_id)
        record = self.env['pb.bank.file'].search(
            [('run_id', '=', run.id), ('state', '=', 'approved')],
            order='id desc', limit=1)
        if not record:
            latest = self.env['pb.bank.file'].search(
                [('run_id', '=', run.id), ('state', '!=', 'superseded')],
                order='id desc', limit=1)
            raise AccessError(_(
                "There is no approved bank file for this pay run yet.%s",
                (_(" The one that was prepared is with %s.",
                   latest._waiting_for() or _('its approver'))
                 if latest else '')))
        return record.action_download()

    @api.model
    def request_release(self, run_id, bank_reference=None):
        """Ask for the money behind the approved file to be sent."""
        self._require_access()
        run = self._run(run_id)
        bank_file = self.env['pb.bank.file'].search(
            [('run_id', '=', run.id), ('state', '=', 'approved')],
            order='id desc', limit=1)
        if not bank_file:
            raise AccessError(_(
                "The bank file has to be approved before the money can be "
                "released."))
        release = self.env['pb.payment.release'].prepare(
            bank_file, bank_reference)
        self.env['biz.approval.engine'].submit(release)
        return self._release_payload(run)

    # ------------------------------------------------------------- payslips out
    @api.model
    def send_payslips(self, run_id, force_all=False):
        self._require_access()
        run = self._run(run_id)
        batch = self._batch(run, create=True)
        batch.action_send(force_all=force_all)
        return self._delivery_payload(batch, run=run)

    # ------------------------------------------------------------- the door
    @api.model
    def open_request(self, request_id):
        """Take the person to the request itself, wherever it is."""
        self._require_access()
        return {
            'type': 'ir.actions.client',
            'tag': 'pb_approval_inbox',
            'name': _('Approvals'),
            'params': {'request_id': int(request_id or 0)},
        }
